Initial commit of idempotent producer.

apurvam opened the pull request:
I think some things to discuss: 

1. Handling of the duplicate sequence number exception. Currently, it is handled transparently on the broker and we return the log append info of the original record. 
2. Handling of the `OutofOrderSequnceException` in the client. 
3. Making the `InitPidRequest` call in the `Sender`. It seems like a lot of boilerplate, because the existing framework is hard to reuse. Is it worth generalizing in this patch, or should we tackle then when we add more RPCs while implementing transactions. 

Todos: 
1. Tests.

apurvam commented:
cc @guozhangwang @hachikuji

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -447,6 +458,11 @@ public boolean hasUnsent() {
             return previous;
     }
 
+    private void closeBatchAndUpdateSequence(RecordBatch batch) {
```
This is a leftover from a previous implementation. Will delete.

apurvam commented:
I added some tests. The changes on the client side are pretty simple, and right now we have unit tests that cover 2 out of the three new client side cases: initPidRequest handling and transaction state handling. We have an integration test that covers the end to end flow. 

What I don't have is a test that explicitly tests the client behavior when it receives an `OutOfOrderSequenceException` from the broker. I looked at the tests in `KafkaProducerTest`, and it appears that we don't have the mocking infra in place to actually simulate particular exceptions encountered when doing `producer.send`. I will investigate what it will take to add these capabilities, as they will be useful when doing transactions as well.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,119 @@
+ * A class which maintains state for transactions. Also keeps the state necessary to ensure idempotent production.
+ */
+public class TransactionState {
+    String transactionalId;
```
This could be `private final` and exposed through a getter (i.e. `transactionId()`)?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,119 @@
+    private short epoch;
+    private boolean isStateValid;
+    private final boolean idempotenceEnabled;
+    Map<TopicPartition, Integer> sequenceNumbers;
```
Same here, `private final`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -244,6 +267,50 @@ public void forceClose() {
+            return;
+        }
+    }
+
```
nit: unneeded newline

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    public boolean pidIsSet() {
+        return pid() != INVALID_PID;
```
nit: any reason to not use the field directly?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,119 @@
+        return idempotenceEnabled;
+    }
+
+    public synchronized void setPid(long pid) {
```
I'm a little bit unclear about the synchronization of this class. Do any threads need to access the transaction state except the sender?

Also, if `pid` is set to anything except `INVALID_PID`, shouldn't we raise an exception? Seems like an illegal state.

apurvam replied:
We don't need this to be synchronized, since `setPid` is only called by the background thread. Also, we should raise an exception if `setPid` is called when `pid` is already set (ie. it is anything but `INVALID_PID`).

However, the transaction state is accessed by multiple threads. In particular, the `sequenceNumber` and `incrementSequenceNumber` are called in the `RecordAccumulator` and `RecordBatch`, which execute in the context of application threads. I think those methods need to be synchronized instead.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -244,6 +267,50 @@ public void forceClose() {
+        metadata.requestUpdate();
+        long start = time.milliseconds();
+        Node node = client.leastLoadedNode(time.milliseconds());
+        while (!client.ready(node, time.milliseconds()) && 0 < remainingTime) {
```
It might make sense to check for disconnects. If the node is disconnected, we can shortcut return and pick another node.

guozhangwang replied:
+1

hachikuji replied:
nit: to be consistent with the check below, should this be `<=`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -244,6 +267,50 @@ public void forceClose() {
+            client.poll(100, time.milliseconds());
+            remainingTime = remainingTime - (time.milliseconds() - start);
+        }
+        if (remainingTime <= 0) {
```
Is this check necessary? We're returning anyway.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -244,6 +267,50 @@ public void forceClose() {
+        InitPidRequest.Builder builder = new InitPidRequest.Builder(null);
+        RequestCompletionHandler callback = new RequestCompletionHandler() {
+            public void onComplete(ClientResponse response) {
+                handleInitPidResponse(response);
```
Doing this through the callback seems unnecessary since we're blocking anyway. Is it just for convenience?

apurvam replied:
It is just for convenience for now. The client used on the consumer has synchronous calls which use callbacks internally. We could unify the two clients in the future, and do away with this call back then.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -163,6 +175,17 @@ public void run() {
         Cluster cluster = metadata.fetch();
+
+        if (transactionState.isIdempotenceEnabled()) {
+            while (!transactionState.pidIsSet() && (time.milliseconds() - now) < maxBlockTimeMs) {
```
I'm not sure about the use of `maxBlockTimeMs` here. My understanding is that this is used to control how long the user will wait in the worst case for `send`, but this is happening in the background. Maybe we should keep trying indefinitely until we've received a PID?

apurvam replied:
This makes sense. To capture our offline discussion: these synchronous background RPC's will try indefinitely. These include the initPidRequest, and all the transactional RPCs. However, the async produce requests will queue behind these synchronous RPC's, and eventually the send will timeout and return an error to the user.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -163,6 +175,17 @@ public void run() {
+            }
+            if (!transactionState.pidIsSet()) {
+                transactionState.setStateInvalid();
+                throw new TimeoutException("Could not get PID within reasonable time.");
```
Note that this exception is caught in `run()` above. I'm not sure that's the intention.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java:
```diff
@@ -204,9 +206,19 @@ public double compressionRate() {
+        if (!isClosed) {
+            recordsBuilder.close();
+            if (transactionState != null) {
+                transactionState.incrementSequenceNumber(topicPartition, this.recordCount);
```
I originally started with having the callers of `close` do this. The major problem however, is that `close` can be called multiple times on the same batch. But we need to ensure that this executes exactly once. By putting this here, guarding by the `isClose` member variable, and making the method synchronized, we can get the exactly once processing in a clear way.

apurvam commented:
I addressed all the comments. Please have a look. Thanks!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java:
```diff
@@ -204,9 +206,19 @@ public double compressionRate() {
+        if (!isClosed) {
+            recordsBuilder.close();
+            if (transactionState != null) {
+                transactionState.incrementSequenceNumber(topicPartition, this.recordCount);
```
Sorry to belabor this, but if we instead expose `recordCount` as a method, then we could let the caller do this. For example, from `RecordAccumulator`, we could do this:

```java
batch.close();
if (transactionState != null)
  transactionState.incrementSequenceNumber(batch.partition(), batch.recordCount());
```
It may not be a huge advantage, but it keeps these classes decoupled.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -244,6 +258,53 @@ public void forceClose() {
+                node = client.leastLoadedNode(time.milliseconds());
+                client.ready(node, time.milliseconds());
+            }
+            client.poll(100, time.milliseconds());
```
Instead of 100, maybe we could use the reconnect timeout? By the way, if we're checking for a disconnect, the time to do it is right after a call to `poll`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -244,6 +258,53 @@ public void forceClose() {
+            client.poll(100, time.milliseconds());
+            remainingTime = remainingTime - (time.milliseconds() - start);
+        }
+        if (remainingTime <= 0) {
```
I may have had a comment about this before, but this check seems pointless.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -244,6 +258,53 @@ public void forceClose() {
+        ClientRequest request = client.newClientRequest(nodeId, builder, time.milliseconds(), true, callback);
+        client.send(request, time.milliseconds());
+        start = time.milliseconds();
+        while (!transactionState.hasPid() && 0 < remainingTime) {
```
If the request fails immediately (say because of a disconnect), it seems like we're still stuck in this loop until expiration of the timeout. I think we need a way to break out if the request fails.

apurvam replied:
So I looked into this. There doesn't seem to be an easy way to handle the case where there is no responseBody. The only way I can see is to add a state variable to the class and check for that in this loop. But that seems like too much complexity for not much benefit. Thoughts?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,129 @@
+        if (!idempotenceEnabled) {
+            throw new IllegalStateException("Attempting to access sequence numbers when idempotence is disabled");
+        }
+        synchronized (sequenceNumbers) {
```
Since there aren't any heavy operations in this class, would it make sense to just synchronize the methods coarsely. It's a lot easier to reason about than having only some state which needs protection.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,129 @@
+        return idempotenceEnabled;
+    }
+
+    public void setPid(long pid) {
```
It might make sense to update the pid and epoch at the same time, so that it is done atomically.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/FileLogInputStream.java:
```diff
@@ -237,6 +237,7 @@ public long checksum() {
             return underlying.checksum();
         }
 
+
```
nit: unintentional?

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -792,6 +798,34 @@ class Log(@volatile var dir: File,
   }
 
+  private def updateAppendInfoForDuplicateEntry(appendInfo: LogAppendInfo) : LogAppendInfo = {
+    val segmentIterator = segments.descendingMap().values().iterator()
```
nit: usually we leave off the parenthesis if the method doesn't mutate state. So `segments.descendingMap.values.iterator` works also.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -792,6 +798,34 @@ class Log(@volatile var dir: File,
 
+  private def updateAppendInfoForDuplicateEntry(appendInfo: LogAppendInfo) : LogAppendInfo = {
+    val segmentIterator = segments.descendingMap().values().iterator()
+    while (segmentIterator.hasNext) {
```
nit: do we need an iterator or can we just use `foreach`?

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -792,6 +798,34 @@ class Log(@volatile var dir: File,
+      val firstSequence = getFirstSequence(segment)
+      if (0 <= firstSequence && firstSequence <= appendInfo.firstSequence) {
+        for (entry <- segment.log.entries.asScala) {
+          if (entry.firstSequence.equals(appendInfo.firstSequence)) {
```
nit: in scala you can use `==`. Below too.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -792,6 +798,34 @@ class Log(@volatile var dir: File,
+    null
+  }
+
+  private def getFirstSequence(segment: LogSegment) : Int = {
```
This method needs to take the PID. Also, we could return `Option` and use `None` instead of -1.

apurvam replied:
Good catch.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -475,28 +474,34 @@ class Log(@volatile var dir: File,
+            flush()
 
+        } catch {
+          case e: DuplicateSequenceNumberException =>
```
Do we need to propagate this error after catching? I'm not sure I see how the error code gets set in the response. Maybe we need to add an `Errors` field to `LogAppendInfo`?

apurvam replied:
I guess we could propagate this. but from the client perspective what value is this information?

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -792,6 +798,34 @@ class Log(@volatile var dir: File,
     deleteOldSegments(shouldDelete)
   }
 
+  private def updateAppendInfoForDuplicateEntry(appendInfo: LogAppendInfo) : LogAppendInfo = {
```
It seems this method doesn't take the requested PID into account. It only checks sequence number.

apurvam replied:
Good catch.

hachikuji commented on core/src/test/scala/integration/kafka/api/ProducerBounceTest.scala:
```diff
@@ -61,9 +63,9 @@ class ProducerBounceTest extends KafkaServerTestHarness {
-    producer1 = TestUtils.createNewProducer(brokerList, acks = 0, bufferSize = producerBufferSize)
-    producer2 = TestUtils.createNewProducer(brokerList, acks = 1, bufferSize = producerBufferSize)
-    producer3 = TestUtils.createNewProducer(brokerList, acks = -1, bufferSize = producerBufferSize)
+    //producer1 = TestUtils.createNewProducer(brokerList, acks = 0, bufferSize = producerBufferSize)
```
Is this intentional?

apurvam replied:
I think I should work multiple producers into the test, since there would be bugs that such a test would catch, as you previous comments suggest :)

hachikuji commented:
By the way, we don't have to do it here, but I think it's definitely worth optimizing for the case that the duplicate sequence number was the last one written to the log. In this case, we can fill the offset and log append time from the pid mapping table instead of doing the expensive seek in the log. We may even consider keeping around n such entries for each PID in order to avoid the seek. I think this is something we'll need to find a reasonable solution for before we can merge into trunk.

apurvam commented:
Updating this PR with our offline discussion. These are the further changes that have to be made to to make this solid: 

1. WE need to enforce max.inflight.requests = 1, Without this setting, it is impossible to guarantee ordering, and we would get `OutOfOrderSequence` routinely when leaders fail.
2. We need to set retries > 0. If retries == 0, then using idempotent producer doesn't make sense.
3. If a batch expires, but was never sent, all subsequent batches for the partition must also be abroted.
4. If a batch expires, and was previously sent, we need to abort all subsequent batches for that partition. Further, we need to reset the producer state by getting a new pid and resetting the partition -> sequenceNumber mapping.

apurvam commented:
One final thing that needs to be done is: If an `OutOfOrderSequence` is encountered. We should raise it to the user. But we should also reset our internal state: get a new Pid, and abort the inflight requests to the bad partition which have the old pid. The inflight requests to other partitions can be flushed. If they fail, we can ignore those responses.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java:
```diff
@@ -159,7 +161,7 @@ else if (this.inRetry() && requestTimeoutMs < (now - (this.lastAttemptMs + retry
         boolean expired = expiryErrorMessage != null;
         if (expired)
-            close();
+            close(null);
```
Should we update the transaction state in this case as well?

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -809,15 +831,25 @@ public boolean isDone() {
+                // This is a fatal exception. All future send / sendOffsets / beginTransaction, endTransaction, etc.
+                // should be aborted with an IllegalStateException.
+                if (transactionState == null) {
+                    throw new IllegalStateException("Received OutOfOrderSequenceException when local state for sequence numbers doesn't exist.");
```
I'd suggest throwing this exception still as an `OutOfOrderSequenceException` or sth. that could be distinguished with other `IllegalStateException`. The reason is that callers, like Streams, need to special handle this case compared with other possibly thrown exceptions. If it is thrown to users as an `IllegalStateException` then it is a bit hard to do and we have to rely on its error message.

apurvam commented:
I did a major revamp of the producer and broker side code. Changes include: 

1. On the broker, we never handled the case where there may be multiple log entries and multiple pids in a single records. This is the case during replication. I updated the code to validate that there are no duplicates even on the follower. If the follower encounters a duplicate, it is treated as a fatal error. 
2. Also on the broker, changed the handling of `DuplicateSequenceException` so that we only permit duplicates of the last appended entry. If there is a duplicate of something in the middle, it is a fatal error. When we encounter a duplicate at the end, we copy the append info from the pid mapping and return that to the user. This is a light weight operation. 
3. When a batch is expired which has been previously sent, we must reset the state of the producer (ie. clear all sequence numbers, and get a new pid).
4. TODO: When a batch is expired which was previously closed (because it was full), and which waited too long to be sent (ie. exceeded `lingerMs`, we need to 'reclaim' it's sequence number or reset the producer as in case 3. This is yet to be done. 
5. When the producer receives an `OutOfOrderSequence` we don't invalidate the producer state. Instead, we reset the PID and sequence numbers, and continue accepting new data.

apurvam commented:
Todos: 

1. Add more tests for the latest code (especially TransactionState synchronization and handling of expired batches).
2. Clean up the implementation of `Sender.getPid` based on PR feedback.
3. Address Guozhang's comments.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -291,14 +294,28 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
+
+            if (idempotenceEnabled) {
+                this.transactionState = new TransactionState(idempotenceEnabled);
+                if (config.getInt(ProducerConfig.RETRIES_CONFIG) == 0) {
```
I think raising an exception is better since it guarantees that the user will see the expectation clearly. Besides, what would we override the value with? MAX_VALUE?

apurvam replied:
I think it's best to throw an exception if the producer isn't configured to be idempotent. Maybe the user doesn't understand the requirements for idempotence, and if they did, may not want to use it? For the initial period especially, I think it is best to set the expectation explicitly.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java:
```diff
@@ -220,6 +220,11 @@
+    public static final String IDEMPOTENCE_ENABLED_CONFIG = "enable.idempotence";
+    public static final String IDEMPOTENCE_ENABLED_DOC = "When set to 'true', the producer will ensure that exactly one copy of each message is written in the stream. If 'false', producer "
+                                                        + "retries due to broker failures, etc., may write duplicates of the retried message in the stream. This is set to 'false' by default. "
+                                                        + "When set to 'true', message loss on the broker will become a fatal exception whose only remedy is the instantiation of a new producer instance.";
```
Guess the second half of this line is no longer true.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -211,17 +214,24 @@ public RecordAppendResult append(TopicPartition tp,
 
+    private MemoryRecordsBuilder getRecordsBuilder(ByteBuffer buffer) {
+        if (transactionState != null) {
+            return MemoryRecords.builder(buffer, LogEntry.MAGIC_VALUE_V2, compression, TimestampType.CREATE_TIME,
```
I think we need to use a version which lets us pass the `batchSize` as we do below.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -257,7 +267,7 @@ private RecordAppendResult tryAppend(long timestamp, byte[] key, byte[] value, C
                             count++;
                             batchIterator.remove();
-                        } else {
+                       } else {
```
nit: misaligned

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -244,6 +272,49 @@ public void forceClose() {
+        Node node = client.leastLoadedNode(time.milliseconds());
+        while (!client.ready(node, time.milliseconds()) && 0 < remainingTime) {
+            if (client.connectionFailed(node)) {
+                node = client.leastLoadedNode(time.milliseconds());
```
We could also just return. The next iteration will select a new node.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -201,9 +215,23 @@ void run(long now) {
+        if (resetState) {
+            transactionState.reset();
+            while (!transactionState.hasPid()) {
+                getPID(requestTimeout);
```
Should we use the retry backoff in between request attempts?

apurvam replied:
What would be the best way to acheive this? `Thread.sleep()`?

apurvam replied:
I've gone with `Thread.sleep` after confirming with Jason that there is no better way in our code at present (ie. no wrapper or other abstraction for such a thing).

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -294,18 +365,33 @@ private void completeBatch(RecordBatch batch, ProduceResponse.PartitionResponse
+                while (!transactionState.hasPid()) {
+                    getPID(requestTimeout);
+                }
+                log.error("The broker received an out of order sequence number for topic-partition {} at offset {}. This indicates data loss on the broker, and should be investigated.",
```
nit: break into two lines

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -294,18 +365,33 @@ private void completeBatch(RecordBatch batch, ProduceResponse.PartitionResponse
+                // batch.
+                this.accumulator.reenqueue(batch, now);
+                this.sensors.recordRetries(batch.topicPartition.topic(), batch.recordCount);
+            }
```
Don't we need to do something in the `else` case? Something like what's done below, completing the batch using `done` and deallocating it.

apurvam replied:
good catch!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -294,18 +365,33 @@ private void completeBatch(RecordBatch batch, ProduceResponse.PartitionResponse
+                // Reset the transactional state and get a new pid, since we can't recover from an out of order
+                // sequence error.
+                transactionState.reset();
+                while (!transactionState.hasPid()) {
```
This is the same snippet we have above. Maybe it should be a method? I'm wondering if we should just add a flag indicating that a PID is needed and then add a check for the flag in the `run` loop? Then we would avoid blocking here before invoking the callback.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,145 @@
+/**
```
We may as well start using the new license header. I had to fix all of these when I submitted the PR for the message format changes.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,145 @@
+     * This method is used when the producer needs to reset it's internal state because of an irrecoverable exception
+     * from the broker.
+    */
+    public synchronized void reset() {
```
Do we need both locks? Could we always use `pidLock`? Otherwise we have to think about cases where we update the sequence number concurrently with a pid reset.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,145 @@
+        }
+    }
+
+    public long pid() {
```
Although `pid` and `epoch` are volatile, we need to ensure that the caller gets the two values atomically. Perhaps we should add a `PidAndEpoch` class or something? Also, maybe we can use `pidLock` to synchronize reads as well. Simple synchronization is usually better.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -163,6 +163,15 @@ else if (maxTimestamp == LogEntry.NO_TIMESTAMP)
+     * @param baseSequence
+     */
+    public void closeWithSequence(int baseSequence) {
+        this.baseSequence = baseSequence;
```
I guess we might want to be a little careful that the record set hasn't already been closed. Perhaps we can raise an `IllegalStateException` if `builtRecords` is not null?

apurvam replied:
This just calls down to `close`, which calls `recordsBuilder.close` which does nothing  if`builtRecords`is not null. Close could be called multiple times already, so I don't think starting throw an `IllegalStateException` is a good idea.

hachikuji replied:
The point is to prevent usage errors. What should the behavior be if you use `closeWithSequence()` and then `close()`. How about the other way around?

apurvam replied:
I see. I think the semantics of `closeWithSequence` do merit an exception if the batch is already closed, since the new sequence number will be ignored. Will update.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/record/MemoryRecordsBuilder.java:
```diff
@@ -453,4 +462,11 @@ public RecordsInfo(long maxTimestamp,
+    /**
+     * Return the ProducerId (Pid) of the LogEntries created by this builder.
+     */
+    public long producerId() {
```
nit: maybe just `pid`?

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -473,29 +473,72 @@ class Log(@volatile var dir: File,
+                    // the duplicate is not at the tail of the log. This should never happen, with a properly written client.
+                    // It should also never happen in the replication case.
+                    // Since this is a error situation throw a fatal exception.
+                    throw new IllegalStateException("Detected a duplicate entry in the middle of the log. This suggests a misconfigured or buggy client. Please see __insert_url__ to understand how clients must be configured when running withe enable.idempotence = true");
```
This would go back to the user as an UNKNOWN error. Maybe we should use `OutOfOrderSequenceException`?

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -473,29 +473,72 @@ class Log(@volatile var dir: File,
+                    throw new IllegalStateException("Detected a duplicate entry in the middle of the log. This suggests a misconfigured or buggy client. Please see __insert_url__ to understand how clients must be configured when running withe enable.idempotence = true");
+                  }
+
+                  if (assignOffsets) {
```
We don't have to do it here, but I feel the replication and produce paths are diverging enough that we ought to just separate them.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -473,29 +473,72 @@ class Log(@volatile var dir: File,
+          // update the PID sequence mapping
+          for ((pid, entry) <- appendInfo.pidEntryMap) {
+            trace(s"Updating pid with sequence: ${pid} -> ${entry._2.seq}")
+            pidMap.update(pid, entry._2.seq, entry._2.epoch, entry._2.offset, entry._2.timestamp)
```
Maybe we can let `update` take a `PidEntry`?

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -793,13 +845,13 @@ class Log(@volatile var dir: File,
 
   /**
-   * The size of the log in bytes
+    * The size of the log in bytes
```
Unintentional?

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -69,10 +72,7 @@ case class LogAppendInfo(var firstOffset: Long,
-                         epoch: Short,
-                         firstSequence: Int,
-                         lastSequence: Int)
+                         pidEntryMap : Map[Long, (PidEntry, PidEntry)])
```
Maybe we could create a case class for the tuple of `PidEntry`. Having names will make it easier to follow the code.

guozhangwang commented:
> Also on the broker, changed the handling of DuplicateSequenceException so that we only permit duplicates of the last appended entry. If there is a duplicate of something in the middle, it is a fatal error. When we encounter a duplicate at the end, we copy the append info from the pid mapping and return that to the user. This is a light weight operation.

We we want to allow duplicates of the last appended entry?

apurvam commented:
> We we want to allow duplicates of the last appended entry?

By 'allow', I mean that we should not consider a duplicate of the last appended entry an error. What we do is return the info (offset and timestamp) of the existing entry.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -448,6 +466,7 @@ private static int parseAcks(String acksString) {
         try {
             // first make sure the metadata for the topic is available
             ClusterAndWaitTime clusterAndWaitTime = waitOnMetadata(record.topic(), record.partition(), maxBlockTimeMs);
+            waitOnPid(maxBlockTimeMs);
```
I think this should be kept separate from `waitOnMetadata`. But yes, we should use the `remainingWaitTime` after `waitOnMetadata` completes, and also update that value after this call is made. Will update the patch.

apurvam commented:
@guozhangwang @hachikuji I addressed all the comments on the PR. Please have another look.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,159 @@
+    }
+
+    boolean hasPid() {
+        return pidAndEpoch.isValid();
```
We can't acquire a lock here, as it would deadlock with the `pidAndEpoch` method, which acquires the lock and then calls `hasPid`. I made pidAndEpoch be volatile and immutable.

apurvam commented:
I found the bug where the integration test would fail intermittently. THe problem was that the broker side assumed valid PIDs start from one. The PidGenerator and client have PIDs starting from 0. So when the broker received a pid of 0, it would do no validation, and the test would detect duplicates and fail. 

I unified the meaning of `INVALID_PID` to be equal to `LogEntry.NO_PID`. Now the checks are consistent everywhere. Part of the problem was that the `LogEntry.NO_PID` was added later through Jason's work and was not ported over to the client code until now. 

Also addressed all the remaining PR comments, and beefed up the `LogTest` cases for duplicate appends. 

I want to beef up the integration test to include more producers with more varied batch sizes. But I think that this is almost done.

apurvam commented:
@hachikuji @guozhangwang please have a look again. I have addressed all comments left on the pr and face to face, and also added lots of unit test cases, and beefed up the integration test. 

Unless there are major comments, I think this is ready to merge at this point.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -422,7 +434,17 @@ public boolean hasUnsent() {
+                                            log.debug("Dest: {} : pid: {}, epoch: {}, Assigning sequence for {}-{} : {}", node,
+                                                    transactionState.pidAndEpoch().pid, transactionState.pidAndEpoch().epoch,
+                                                    batch.topicPartition.topic(), batch.topicPartition.partition(),
+                                                    transactionState.sequenceNumber(batch.topicPartition));
```
nit: we can call `sequenceNumber()` once and cache the value locally, since the function itself is synchronized an hence may be expensive. Similarly for `pidAndEpoch()`.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java:
```diff
@@ -422,7 +434,17 @@ public boolean hasUnsent() {
+                                                    transactionState.pidAndEpoch().epoch,
+                                                    transactionState.sequenceNumber(batch.topicPartition));
+                                        } else {
+                                            batch.close();
```
`close()` now only get called in `abortBatches`. I'm wondering if we could collapse these two with `close(TransactionState)` in which we will update the state if it is not null?

apurvam replied:
See Jason's comment below. I added a `setProducerState` method, and removed the `closeWithTransactionState`. This way, we have only one close path, with the producer state being set only on a successful send if idempotence is enabled.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java:
```diff
@@ -206,11 +209,21 @@ public double compressionRate() {
 
     public void close() {
         recordsBuilder.close();
+        isWritable = false;
```
See my comment above.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java:
```diff
@@ -206,11 +209,21 @@ public double compressionRate() {
+
+    public void closeWithTransactionState(long pid, short epoch, int baseSequence) {
+        recordsBuilder.closeWithTransactionState(pid, epoch, baseSequence);
+        isWritable = false;
```
Not sure why we need this additional variable. Could we just modify the logic of `isWritable` below to check that the builder is not close AND still have space then, which seems simpler?

guozhangwang replied:
Just explaining my observations here: the only place we were not closing the builder while setting it to false at the same time is line 80 above. And if we want to reflect that a previous tryAppend has failed in `isWritable` we close just close the batch as we did before after `tryAppend` failed.

apurvam replied:
The problem is that we write the headers (sequence number, pid, etc.), on close. So the only time we can do a successful close (ie when we want to actually send the batch), is at the moment before we send it (ie. in `drain`). That is why this new variable was introduced: we can't simply close the batch when there is no further room, etc. because then we would have to rewrite the headers later if we actually wanted to send the batch. 

The choice is between this variable (which blocks further writes _without_ closing), and the rewriting the final record everytime 'close' is called.. But the latter will _always_ happen on a successful send, so this seems like the better approach.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,6 +284,65 @@ public void forceClose() {
+    }
+
+    private void maybeWaitForPid() {
+        while (!transactionState.hasPid()) {
```
We have a nested "while" loop here and I'm wondering if it is better to flatten in into a single while loop. Take `AbstractCoordinator.ensureCoordinatorReady` as an example, which does the following:

```
while (not found pid and not timed out) {

  if( findANode) {
     send()
     blockWaitForResponse
  } else {
     // either throw an exception, or continue to the next loop to retry
  }

  // back of if needed
}

apurvam replied:
This is a good piece of advice. I have incorporated this pattern. It is much cleaner, thanks!

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -580,20 +595,33 @@ private ClusterAndWaitTime waitOnMetadata(String topic, Integer partition, long
     }
 
+    private long waitOnPid(long maxWaitMs) throws InterruptedException {
+        if (transactionState == null) {
```
Either that or use `maybeWaitOnPid` to emphasize the fact that it doesn't necessarily involve waiting. Both options work for me.

apurvam replied:
I'll change this to `maybeWaitOnPid`. Would prefer to isolate all the logic into this function, since it's called multiple times.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -473,29 +474,88 @@ class Log(@volatile var dir: File,
+              }
+          }
+        }
+        // If this is a follower and there was a duplicate in the incoming record, a `DuplicateSequenceException` would be thrown and
```
Do we need this comment? Seems like the code above is clear enough.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -460,6 +474,7 @@ private static int parseAcks(String acksString) {
             // first make sure the metadata for the topic is available
             ClusterAndWaitTime clusterAndWaitTime = waitOnMetadata(record.topic(), record.partition(), maxBlockTimeMs);
             long remainingWaitMs = Math.max(0, maxBlockTimeMs - clusterAndWaitTime.waitedOnMetadataMs);
+            remainingWaitMs = waitOnPid(remainingWaitMs);
```
It is passed on to the `accumulator.append`. This line just reflects that we spend more of our block time budget on waiting for the pid.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -210,9 +229,21 @@ void run(long now) {
+
+        if (resetState) {
+            transactionState.reset();
+            maybeWaitForPid();
```
discussed off line. It makes sense to terminate the loop at this point.

apurvam commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -24,13 +24,28 @@ import java.nio.file.Files
+    lastOffset - numRecords + 1
+  }
+}
+private[log] case class PidEntryTuple(firstEntry: PidEntry, lastEntry: PidEntry)
```
Will change to `PidEntryRange`. But I would prefer leaving this as `firstEntry` and `lastEntry`. They read better in actual usage. These names were `first` and `last`, but they made the validation code harder to read, IMO.

apurvam commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,6 +284,65 @@ public void forceClose() {
+            return;
+        }
+
+        final AtomicBoolean initPidRequestCompleted = new AtomicBoolean(false);
```
I would prefer to do this as part of that PR, when the actual use is clear.

apurvam commented:
@hachikuji @guozhangwang I think I addressed all the PR comments. Please take a look.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java:
```diff
@@ -296,6 +296,21 @@ private KafkaProducer(ProducerConfig config, Serializer<K> keySerializer, Serial
+                this.transactionState = new TransactionState(idempotenceEnabled, time);
+                if (config.getInt(ProducerConfig.RETRIES_CONFIG) == 0) {
+                    throw new ConfigException("Need to set '" + ProducerConfig.RETRIES_CONFIG
+                            + "' to greater than zero in order to use the idempotent producer.");
```
Might be helpful to offer a brief explanation. For example, "The producer uses sequence numbers to deduplicate messages on the broker; without retries, no duplication is possible in the first place."

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -210,9 +229,21 @@ void run(long now) {
-        // update sensors
-        for (RecordBatch expiredBatch : expiredBatches)
+
+        // Reset the PID if an expired batch has previously been sent to the broker. Aloso update the metrcis
```
typo: "Aloso" and "metrcis"

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,6 +284,58 @@ public void forceClose() {
+        client.send(request, time.milliseconds());
+    }
+
+    private Node getReadyNode(long remainingTimeMs) {
```
Take a look at `NetworkClientBlockingOps.blockingReady()` since this is super tricky to get right. It definitely needs to be a loop because `poll()` may return for reasons other than the event we are looking for. The call to `connectionFailed` also needs to be done after the call to `poll()` since `ready()` will reset the state.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,6 +284,58 @@ public void forceClose() {
+                Node node = getReadyNode(requestTimeout);
+                if (node == null) {
+                    log.warn("Could not find an available node while trying to get a producer id.");
+                    throw new NoAvailableBrokersException();
```
Maybe we should simply backoff (i.e. sleep for the retry backoff duration) and try again since this is the background thread and we cannot propagate exceptions. I think we usually expect this to be a transient state anyway (we assume a broker will eventually come online).

apurvam replied:
We do backoff, this exception is caught and logged, at which point we backoff and retry.

hachikuji replied:
Got it. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,6 +284,58 @@ public void forceClose() {
+        String nodeId = node.idString();
+        InitPidRequest.Builder builder = new InitPidRequest.Builder(null);
+        RequestCompletionHandler callback = new RequestCompletionHandler() {
+            public void onComplete(ClientResponse response) {
```
Did we lose the atomic boolean we were using to tell when this had returned?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -253,6 +284,58 @@ public void forceClose() {
+                    log.warn("Could not find an available node while trying to get a producer id.");
+                    throw new NoAvailableBrokersException();
+                }
+                sendInitPidRequest(node);
```
I think we need to wait for this to return. The completion of the `poll()` below does not necessarily indicate that the request returned.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -303,21 +386,42 @@ private void completeBatch(RecordBatch batch, ProduceResponse.PartitionResponse
+                this.accumulator.reenqueue(batch, now);
+                this.sensors.recordRetries(batch.topicPartition.topic(), batch.recordCount);
+            } else {
+                // TODO(apurva): Is 'KafkaException' the best one to throw here? What would be more appropriate?
```
Maybe we should treat this as an `OutOfSequenceException`. That is the error which caused us to get a new pid and probably what would happen if we tried to resend this batch.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionState.java:
```diff
@@ -0,0 +1,129 @@
+ */
+public class TransactionState {
+    private volatile PidAndEpoch pidAndEpoch;
+    private final boolean idempotenceEnabled;
```
We don't need this anymore because we let `TransactionState` be null if idempotence is disabled.

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -504,6 +516,58 @@ class Log(@volatile var dir: File,
+          }
+      }
+    }
+    return isDuplicate
```
nit: no need for `return`

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -504,6 +516,58 @@ class Log(@volatile var dir: File,
+
+  def validatePidEntries(appendInfo: LogAppendInfo, assignOffsets: Boolean): Boolean = {
+    var isDuplicate = false
+    for ((pid, incomingRecord) <- appendInfo.pidEntryMap) {
```
nit: `incomingRecord` -> `incomingEntries`?

hachikuji commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -530,29 +594,34 @@ class Log(@volatile var dir: File,
+      }
 
-      // update the last offset seen
+     // update the last offset seen
```
nit: misaligned

hachikuji commented on core/src/main/scala/kafka/log/ProducerIdMapping.scala:
```diff
@@ -24,13 +24,28 @@ import java.nio.file.Files
 import scala.collection.mutable
 
-private[log] case class PidEntry(seq: Int, epoch: Short, offset: Long)
+case class PidEntry(lastSeq: Int,
```
This should still be `private[log]`. Probably the reason it doesn't work is that `ProducerIdMappingTest` is in the wrong package. Instead of `unit.kafka.idempotence`, it should be `kafka.log`.

apurvam commented:
@hachikuji : I ported the `NetworkClientBlockingOps` methods to a new `BlockingNetworkClient` class. This takes care of the mechanics of doing blocking calls, and can be reused for the transactional RPCs later as well.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -109,6 +112,9 @@
     private final TransactionState transactionState;
 
+    /* A blocking version of the network client to be used for the new blocking RPCs for the idempotent producer */
+    private final BlockingNetworkClient blockingNetworkClient;
```
It seems a little confusing to have two references to the same underlying NetworkClient. Maybe we should make this `NetworkClientUtils` or something like that and just implement the methods as static utility functions? Unfortunately java doesn't have great mixin patterns.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java:
```diff
@@ -318,21 +314,19 @@ private void maybeWaitForPid() {
-                sendInitPidRequest(node);
-                client.poll(requestTimeout, time.milliseconds());
+                ClientResponse response = sendInitPidRequest(node);
+                if (response.hasResponse() && (response.responseBody() instanceof  InitPidResponse)) {
```
nit: extra space after `instanceof`

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/BlockingNetworkClient.java:
```diff
@@ -0,0 +1,109 @@
+     * This method is useful for implementing blocking behaviour on top of the non-blocking `NetworkClient`, use it with
+     * care.
+     */
+    public boolean isReady(Node node, Time time, long timeout) throws IOException {
```
Maybe `awaitReady` would be clearer? `isReady` doesn't suggest that any blocking is involved.

hachikuji replied:
One other minor nit: can we rename `timeout` to `timeoutMs` to make the units clear?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/BlockingNetworkClient.java:
```diff
@@ -0,0 +1,109 @@
+
+        long attemptStartTime = startTime;
+
+        if (!isReady(node, time.milliseconds()))
```
We seem to make a lot of unnecessary calls to `time.milliseconds()`. We have actually had performance bugs around this call, so where reasonable, we try to reuse the results. For this line and the one below, we could probably just reuse `startTime`. Similarly inside the loop below.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/BlockingNetworkClient.java:
```diff
@@ -0,0 +1,109 @@
+            for (ClientResponse response : responses) {
+                if (response.requestHeader().correlationId() == request.correlationId()) {
+                    if (response.wasDisconnected()) {
+                        throw new IOException("Connection to {} " + response.destination() + " was disconnected before ther response was read");
```
typo: "ther"

ijuma commented on checkstyle/import-control.xml:
```diff
@@ -98,6 +98,8 @@
       <allow pkg="org.apache.kafka.common.protocol.types" />
       <allow pkg="org.apache.kafka.common.errors" />
+      <!-- MemoryRecordsBuilder needs to access TransactionState, which seems to naturally belong in the Producer -->
+      <allow pkg="org.apache.kafka.clients.producer" />
```
Package `common` should not depend on package `client`.

apurvam replied:
Yes. That is awkward. The dependence on the new `TransactionState` class is very minimal. I refactored to remove the dependency: https://github.com/confluentinc/kafka/pull/143
